File: elixir/food-chain/lib/food_chain.ex
defmodule FoodChain do
  @tiere ~w(fly spider bird cat dog goat cow horse)
  @ausrufe [
    "I don't know why she swallowed the fly. Perhaps she'll die.",
    "It wriggled and jiggled and tickled inside her.",
    "How absurd to swallow a bird!",
    "Imagine that, to swallow a cat!",
    "What a hog, to swallow a dog!",
    "Just opened her throat and swallowed a goat!",
    "I don't know how she swallowed a cow!",
    "She's dead, of course!"
  ]

  defp sonderfall(n) do
    if @tiere |> Enum.at(n) == "spider", do: " that wriggled and jiggled and tickled inside her", else: ""
  end

  defp verse(n) do
    erste_zeilen = ["I know an old lady who swallowed a #{@tiere |> Enum.at(n-1)}.", @ausrufe |> Enum.at(n-1)]
    if n == 8 or n == 1 do
      (erste_zeilen |> Enum.join("\n")) <> "\n"
    else
      mittlere_zeilen =
        for i <- Enum.reverse(1..n-1) do
          "She swallowed the #{@tiere |> Enum.at(i)} to catch the #{@tiere |> Enum.at(i-1)}#{sonderfall(i-1)}."
        end
      letzte_zeile = [@ausrufe |> Enum.at(0)]
      (erste_zeilen ++ mittlere_zeilen ++ letzte_zeile |> Enum.join("\n")) <> "\n"
    end
  end

  @doc """
  Generate consecutive verses of the song 'I Know an Old Lady Who Swallowed a Fly'.
  """
  @spec recite(start :: integer, stop :: integer) :: String.t()
  def recite(start, stop) do
    start..stop |> Enum.map(&verse/1) |> Enum.join("\n")
  end
end

File: fsharp/food-chain/FoodChain.fs
module FoodChain
open System

let tiere = ["fly"; "spider"; "bird"; "cat"; "dog"; "goat"; "cow"; "horse"]
let ausrufe = [ "I don't know why she swallowed the fly. Perhaps she'll die.";
         "It wriggled and jiggled and tickled inside her.";
         "How absurd to swallow a bird!";
         "Imagine that, to swallow a cat!";
         "What a hog, to swallow a dog!";
         "Just opened her throat and swallowed a goat!";
         "I don't know how she swallowed a cow!";
         "She's dead, of course!"]

let sonderfall n : string =
    if tiere[n - 1] = "spider" then " that wriggled and jiggled and tickled inside her" else ""

let verse n : string list =
    let ersteZeilen = [$"I know an old lady who swallowed a {tiere[n-1]}."; $"{ausrufe[n-1]}"]
    if n = 8 || n = 1 then ersteZeilen
    else 
        ersteZeilen @
        ([n-1..-1..1] 
        |> List.map (fun i -> $"She swallowed the {tiere[i]} to catch the {tiere[i - 1]}{sonderfall(i)}.")) 
        @ [$"{ausrufe[0]}"]

let recite start stop = ([start..stop - 1] |> List.map verse |> List.map (fun ls -> ls @ [""]) |> List.concat) @ verse stop